students/tlugosan/lesson06/test_mailroom4.py

```diff
@@ -0,0 +1,121 @@
+        captured = capsys.readouterr()
+        out = captured.out.strip()
+        err = captured.err.strip()
+        assert ('Dear {}, Thank you for your generous contribution of ${:.2f} to our program.'.format(test_input[2],
+                                                                                                      test_input[
+                                                                                                          3])) in out
+def test_thanks_new_donor(capsys):
+    test_input = ['1', 'list', "Bruno Mars", 450, '4']
+    with mock.patch('builtins.input', side_effect=test_input):
+        mailroom.select_action_dictionary('1', mailroom.switch_func_dict)
+        captured = capsys.readouterr()
+        out = captured.out.strip('\n')
+        err = captured.err.strip()
+        assert 'Dear {}, Thank you for your generous contribution of ${:.2f} to our program.'.format(test_input[2],
+                                                                                                     test_input[
+                                                                                                         3]) in out
+
+
+def test_thanks_negative_donation(capsys):
+    test_input = ['1', 'list', "Toni Orlando", '-4', '1', '4']
+    with mock.patch('builtins.input', side_effect=test_input):
+        mailroom.select_action_dictionary('1', mailroom.switch_func_dict)
+        captured = capsys.readouterr()
+        out = captured.out.strip('\n')
+        err = captured.err.strip()
+        assert 'Number has to be positive.' in out
+
+
+def test_thanks_not_numeric_donation(capsys):
+    test_input = ['1', 'list', "Toni Orlando", 'ab', '1', '4']
+    with mock.patch('builtins.input', side_effect=test_input):
+        mailroom.select_action_dictionary('1', mailroom.switch_func_dict)
+        captured = capsys.readouterr()
+        out = captured.out.strip('\n')
+        err = captured.err.strip()
+        assert 'Input must be a number.' in out
+
+
+def test_thanks_directly(capsys):
+    test_input = ["Marco Polo", 100]
+    with mock.patch('builtins.input', side_effect=test_input):
+        mailroom.sending_thank_you()
+        captured = capsys.readouterr()
+        out = captured.out.strip('\n')
+        err = captured.err.strip()
+        assert 'Dear {}, Thank you for your generous contribution of ${:.2f} to our program.'.format(*test_input) == out
+
+
+def test_print_report(capsys):
+    result_list = mailroom.print_report()
+    captured = capsys.readouterr()
+    out = captured.out.strip()
+    test_list = []
+    for k in mailroom.table_dictionary:
+        sum_donations = sum(mailroom.table_dictionary[k])
+        total_gifts = len(mailroom.table_dictionary[k])
+        average_gift = sum_donations / total_gifts
+        test_list.append([k, sum_donations, total_gifts, average_gift])
+        test_sorted_new_list = sorted(test_list, key=itemgetter(1), reverse=True)
```

Comment:
This resembles your actual mailroom code - which tells me that this isn't really testing your feature code?
if there is a bug in this code in your mailroom, your test wouldn't catch it because it's the same code here.

It is best for your tests to be a little more explicit, like having expected be a report string or list of formatted string rows. This would require your print_report to be refactored into logic function and then separate print function.